Sources/MongoSwift/MongoDatabase.swift

```diff
@@ -223,30 +212,64 @@ public class MongoDatabase {
      *   - options: Optional `ListCollectionsOptions` to use when executing this command
      *   - session: Optional `ClientSession` to use when executing this command
      *
-     * - Returns: a `MongoCursor` over an array of collections
+     * - Returns: a `MongoCursor` over an array of `CollectionSpecification`s
      *
      * - Throws:
      *   - `userError.invalidArgumentError` if the options passed are an invalid combination.
      *   - `UserError.logicError` if the provided session is inactive.
      */
     public func listCollections(_ filter: Document? = nil,
                                 options: ListCollectionsOptions? = nil,
-                                session: ClientSession? = nil) throws -> MongoCursor<Document> {
-        var opts = try encodeOptions(options: options, session: session)
-        if let filterDoc = filter {
-            opts = opts ?? Document()
-            // swiftlint:disable:next force_unwrapping
-            opts!["filter"] = filterDoc // guaranteed safe because of nil coalescing default.
+                                session: ClientSession? = nil) throws -> MongoCursor<CollectionSpecification> {
+        let operation = ListCollectionsOperation(database: self, filter: filter, options: options, nameOnly: false)
+        guard case let .specs(result) = try self._client.executeOperation(operation, session: session) else {
+            throw RuntimeError.internalError(message: "Invalid result")
         }
+        return result
+    }
 
-        return try MongoCursor(client: self._client, decoder: self.decoder, session: session) { conn in
-            self.withMongocDatabase(from: conn) { dbPtr in
-                guard let collections = mongoc_database_find_collections_with_opts(dbPtr, opts?._bson) else {
-                    fatalError(failedToRetrieveCursorMessage)
-                }
-                return collections
-            }
+    /**
+     * Gets a list of `MongoCollection`s in this database.
+     *
+     * - Parameters:
+     *   - filter: a `Document`, optional criteria to filter results by
+     *   - options: Optional `ListCollectionsOptions` to use when executing this command
+     *   - session: Optional `ClientSession` to use when executing this command
+     *
+     * - Returns: An array of `MongoCollection`s that match the provided filter.
+     *
+     * - Throws:
+     *   - `userError.invalidArgumentError` if the options passed are an invalid combination.
+     *   - `UserError.logicError` if the provided session is inactive.
+     */
+    public func listMongoCollections(_ filter: Document? = nil,
+                                     options: ListCollectionsOptions? = nil,
+                                     session: ClientSession? = nil) throws -> [MongoCollection<Document>] {
+        return try self.listCollectionNames(filter, options: options, session: session).map { self.collection($0) }
```

Comment:
From the [section on "returning a list of collection objects"](https://github.com/mongodb/specifications/blob/master/source/enumerate-collections.rst#returning-a-list-of-collection-objects): 
>Drivers MUST specify the nameOnly option when executing the listCollections command for this method, except drivers MUST NOT set nameOnly if a filter specifies any keys other than name.

So it looks like we need to inspect the `filter` (if there is one) to see what we should set `nameOnly` to. The reason being that there are big performance differences to using `nameOnly: true` vs `nameOnly: false`, and we want to take advantage of them where possible.